polyhook2/Virtuals/VTableSwapHook.hpp

```diff
@@ -40,5 +41,64 @@ class VTableSwapHook : public PLH::IHook {
 	VFuncMap m_origVFuncs;
 	bool m_Hooked;
 };
+
+
+// storage class for address of a virtual function
+// also stores the function pointer type and index number on the class level
+template<uint16_t I, typename FuncPtr>
+struct VFunc {
+	VFunc() : func(nullptr) {};
+	VFunc(FuncPtr f) : func(f) {};
+	const FuncPtr func;
+	static const uint16_t func_index;
+	typedef FuncPtr func_type;
+};
+
+// definition of constant must reside outside class declaration
+template<uint16_t I, typename FuncPtr> const uint16_t VFunc<I, FuncPtr>::func_index = I;
+
+namespace detail {
+
+	// helper function to convert sequence of VFunc structs into a VFuncMap
+	// using recursive template definition
+	inline PLH::VFuncMap make_vfunc_map()
+	{
+		return PLH::VFuncMap{ };
+	};
+
+	template<uint16_t I, typename FuncPtr, typename ... VFuncTypes>
+	inline PLH::VFuncMap make_vfunc_map(VFunc<I, FuncPtr> vfunc, VFuncTypes ... vfuncs)
+	{
+		PLH::VFuncMap map{ {I, reinterpret_cast<uint64_t>(vfunc.func)} };
+		map.merge(make_vfunc_map(vfuncs ...));
+		return map;
+	};
+
+}
+
+template<class ClassType, typename ... VFuncTypes>
+class VTableSwapHook2 : private VTableSwapHook {
```

Comment:
Please name this something like VTableSwapHookHelper or something more descriptive.

Reply:
I've currently removed the helper class, however I can resurrect it if you like. It could provide some extra type safety when calling origFunc, i.e. ensuring that the VFunc template argument is one of the VFuncTypes through a static_assert, besides just hooking and unhooking automatically on constructor/destructor. I guess it could also carry a shared_ptr to the original class to ensure the original does not get destroyed before the hook is destroyed. Let me know if you feel this functionality would be useful in a helper class for the library.